File: Leetcode/147-Insertion-Sort-List/InsertionSortList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	dummy := &ListNode{Val: head.Val}
	for head != nil {
		temp := head
		head = head.Next
		temp.Next = nil
		n := dummy
		for n.Next != nil {
			if temp.Val < n.Next.Val {
				break
			}
			n = n.Next
		}
		temp.Next = n.Next
		n.Next = temp
	}
	return dummy.Next
}

File: Leetcode/147-Insertion-Sort-List/InsertionSortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode* dummy = new ListNode(head->val);
        ListNode* temp;
        while (head!=NULL) {
            temp = head;
            head = head->next;
            temp->next = NULL;
            ListNode* n = dummy;
            while (n->next!=NULL) {
                if (temp->val < n->next->val) {
                    break;
                }
                n = n->next;
            }
            temp->next = n->next;
            n->next = temp;
        }
        return dummy->next;
    }
};
